nifi-nar-bundles/nifi-asn1-bundle/nifi-asn1-services/src/main/java/org/apache/nifi/jasn1/convert/converters/TbcdStringConverter.java

```diff
@@ -0,0 +1,110 @@
+ */
+package org.apache.nifi.jasn1.convert.converters;
+
+import com.beanit.asn1bean.ber.types.BerOctetString;
+import com.beanit.asn1bean.ber.types.BerType;
+import org.apache.nifi.jasn1.convert.JASN1Converter;
+import org.apache.nifi.jasn1.convert.JASN1TypeAndValueConverter;
+import org.apache.nifi.serialization.record.DataType;
+import org.apache.nifi.serialization.record.RecordFieldType;
+
+public class TbcdStringConverter implements JASN1TypeAndValueConverter {
+
+    private static final String TBCD_STRING_TYPE = "TBCDSTRING";
+    private static final char[] TBCD_SYMBOLS = "0123456789*#abc".toCharArray();
+
+    @Override
+    public boolean supportsType(Class<?> berType) {
+        boolean supportsType = BerOctetString.class.isAssignableFrom(berType) && isTbcdString(berType);
+
+        return supportsType;
+    }
+
+    @Override
+    public DataType convertType(Class<?> berType, JASN1Converter converter) {
+        DataType dataType = RecordFieldType.STRING.getDataType();
+
+        return dataType;
+    }
+
+    @Override
+    public boolean supportsValue(BerType value, DataType dataType) {
+        boolean supportsValue = value instanceof BerOctetString && isTbcdString(value.getClass());
+
+        return supportsValue;
+    }
+
+    @Override
+    public Object convertValue(BerType value, DataType dataType, JASN1Converter converter) {
+        final BerOctetString berValue = ((BerOctetString) value);
+
+        byte[] bytes = berValue.value;
+
+        int size = (bytes == null ? 0 : bytes.length);
+        StringBuilder resultBuilder = new StringBuilder(2 * size);
+
+        for (int octetIndex = 0; octetIndex < size; ++octetIndex) {
+            int octet = bytes[octetIndex];
+
+            int digit2 = (octet >> 4) & 0xF;
+            int digit1 = octet & 0xF;
+
+            if (digit1 == 15) {
+                invalidFiller(octetIndex, octet);
+            } else if (digit1 > 15) {
+                invalidInteger(digit1);
+            } else {
+                resultBuilder.append(TBCD_SYMBOLS[digit1]);
+            }
+
+            if (digit2 == 15) {
+                if (octetIndex != size - 1) {
+                    invalidFiller(octetIndex, octet);
+                }
+            } else if (digit2 > 15) {
+                invalidInteger(digit2);
+            } else {
+                resultBuilder.append(TBCD_SYMBOLS[digit2]);
+            }
+        }
+
+        return resultBuilder.toString();
+    }
+
+    private boolean isTbcdString(Class<?> berType) {
+        Class<?> currentType = berType;
+        while (currentType != null) {
+            if (currentType.getSimpleName().equals(TBCD_STRING_TYPE)) {
+                return true;
+            }
+
+            currentType = currentType.getSuperclass();
+        }
+
+        return false;
+    }
+
+    private void invalidFiller(int octetIndex, int octet) {
+        throw new NumberFormatException("Illegal filler in octet nr. " + octetIndex + ": " + octet);
```

Comment:
Recommend spelling out the message:
```suggestion
        throw new NumberFormatException("Illegal filler in octet number [" + octetIndex + "] " + octet);
```

Reply:
Curiously, I think the original notation is actually less ambiguous.

The term `octet number [15]` could easily be interpreted as an expression that states an octet _value_ was found to be 15 and was considered to be illegal (`number` having the ambiguous meaning here).

`octet nr. 15` implies better that we are talking about the 15th octet.

I understand that after that we write the value as well, so really it would look something like this: `octet number [15] 12` but it would still be very confusing in my opinion.
The original in turn would look like `octet nr. 15: 12`.

I can add quotations or brackets around the _value_ like this: `octet nr. 15: '12'` or this: `octet nr. 15: [12]` but not convinced it would be actually better. I would add them for strings but for numbers I think we are better off without them.

Reply:
Thanks for the reply and additional background. I'm not too concerned about the formatting, just the use of `nr.` does not seem as clear as it could be. What about `octet index`? Or perhaps just dropping it altogether and using something like: `octet 22: 15` or `octet 22: value 15`. Just looking to avoid the use of `nr.` for clarity.

Reply:
I think "nr." is a fairly accepted term in such a context but I can remove it.